Unit tests

Charana123 opened the pull request:
instructions - 

    download packages - npm install
    install grunt globally (only done one) - npm install grunt
    transpile javascript - grunt
    run tests - npm test

dougszumski commented on spec/monasca_client_tests.js:
```diff
@@ -0,0 +1,26 @@
+import nock from 'nock';
+
+export class monasca_api_mock {
```
Is this left over from testing? It doesn't look like we use it? Can we delete it?

dougszumski commented on src/components/alarm_definitions.js:
```diff
@@ -22,9 +22,9 @@ import MonascaClient from './monasca_client';
 
   /** @ngInject */
-  constructor($scope, $injector, backendSrv, datasourceSrv, alertSrv) {
+  constructor($scope, $injector, backendSrv, datasourceSrv, alertSrv, monascaClientSrv) {
```
Can we avoid modifying the code that we're testing?

Charana123 replied:
you can't test the grafana components without dependency injection in the constructor

dougszumski commented on src/components/edit_alarm_definition.js:
```diff
@@ -73,13 +73,13 @@ export class EditAlarmDefinitionPageCtrl {
       .then(alarm_definition => {
-	this.savedAlarmDefinition = this.pickKnownFields(alarm_definition);
-	this.newAlarmDefinition = _.cloneDeep(this.savedAlarmDefinition);
+	      this.savedAlarmDefinition = this.pickKnownFields(alarm_definition);
```
Please put all re-formatting commits in a separate commit. The idea is that it makes the actual changes to the code easier to pick out.

dougszumski commented on src/components/monasca_client.js:
```diff
@@ -235,59 +234,62 @@ export default class MonascaClient {
-	  .catch(err => { throw err; });
-      })
-      .catch(err => { throw err; });
+    .then(response => {
```
Please put this reformatting change in a separate commit.

dougszumski replied:
This still needs to be moved to the reformat commit.

dougszumski reviewed (changes requested):
Thanks @Charana123! A couple of nits to fix before we get started..

dougszumski commented on src/components/config.js:
```diff
@@ -21,7 +21,7 @@ export class MonascaAppConfigCtrl {
     }
 
-    var show_types = ['monasca-grafana-datasource', 'monasca-datasource'];
+    var show_types = ['monasca-grafana-datasource', 'monasca-datasource', 'test-'];
```
I don't think this should be in the commit?

dougszumski commented on src/components/edit_alarm_definition.js:
```diff
@@ -22,10 +22,10 @@ import _ from 'lodash';
 
   /** @ngInject */
-  constructor($scope, $injector, $location, backendSrv, datasourceSrv, alertSrv) {
+  constructor($scope, $injector, $location, backendSrv, datasourceSrv, alertSrv, monascaClientSrv) {
```
Can we remove backendSrv, datasourceSrv now that you're injecting the monasca client? They don't appear to be used?

brtknr replied:
+1

dougszumski commented on src/components/edit_notification.js:
```diff
@@ -21,10 +21,10 @@ import MonascaClient from './monasca_client';
 
   /** @ngInject */
-  constructor($scope, $injector, $location, backendSrv, datasourceSrv, alertSrv) {
+  constructor($scope, $injector, $location, backendSrv, datasourceSrv, alertSrv, monascaClientSrv) {
```
Can we remove backendSrv, datasourceSrv now that you're injecting the monasca client? They don't appear to be used?
(Same for other instances)

dougszumski commented on src/components/monasca_client.js:
```diff
@@ -235,59 +234,62 @@ export default class MonascaClient {
-      });
-    }).catch(err => { throw err; });
+    return Promise.reject( { message: 'MY ERROR!'} )
+  //   return this._getDataSource().then(data_source => {
```
I guess commenting this out isn't supposed to be in the commit?

If you haven't figured it out already (I'm sure you have :) ) you can use ```git rebase -i <commit_id>``` to go back and amend previous commits.

brtknr commented on src/components/monasca_client.js:
```diff
@@ -15,8 +15,8 @@
 
-import config from 'app/core/config';
-import appEvents from 'app/core/app_events';
+// import config from 'app/core/config';
```
Remove these if not needed

brtknr commented on src/components/monasca_client.js:
```diff
@@ -15,8 +15,8 @@
-import config from 'app/core/config';
-import appEvents from 'app/core/app_events';
+// import config from 'app/core/config';
+// import appEvents from 'app/core/app_events';
```
This too

brtknr commented on src/spec/monasca_client_spec.js:
```diff
@@ -0,0 +1,78 @@
+
+function datasourceSrvMock(){
+
+}
```
Will backendSrvMock and datasourceSrvMock be necessary after removing them as inputs?

brtknr commented:
> download packages - npm install
> install grunt globally (only done one) - npm install grunt
> transpile javascript - grunt
> run tests - npm test

Add these to README

Charana123 committed:
monasca_client constructor DI

configured `MonascaClient` as a angular service for constructor DI
allowing mocking of the service for unit-testing

matrixik commented:
Could you use `Start a review` instead of `Add single comment`?
This way github won't be sending so many mails...

dougszumski commented on src/components/monasca_client.js:
```diff
@@ -20,14 +20,14 @@ import appEvents from 'app/core/app_events';
 
 export default class MonascaClient {
 
+  /** @ngInject */
```
Please move this out of the reformat commit.

dougszumski commented on src/components/monasca_client.js:
```diff
@@ -233,59 +228,60 @@ export default class MonascaClient {
-	'Content-Type': 'application/json',
-	'X-Auth-Token': data_source.token
+	      'Content-Type': 'application/json',
+	      'X-Auth-Token': data_source.token
```
Should be in the reformat commit.

dougszumski commented on src/components/monasca_client.js:
```diff
@@ -233,59 +228,60 @@ export default class MonascaClient {
-	      monasca_response = err.data.description;
-	    } else{
+	          }
+	          else if (err.data.description) {
```
Same for this..

dougszumski commented on Gruntfile.js:
```diff
@@ -64,3 +64,6 @@ module.exports = function(grunt) {
 
   grunt.registerTask('default', ['clean', 'sass', 'copy', 'babel']);
 };
+
```
Please remove - it's good practice to keep change requests as small as possible.

dougszumski commented on karma.conf.js:
```diff
@@ -69,8 +73,10 @@ module.exports = function(config) {
-    browsers: ['Chrome'],
-
+    browsers: [
+      // 'Chrome'
```
I think we can delete this line.

dougszumski commented on src/spec/monasca_client.spec.js:
```diff
@@ -0,0 +1,14 @@
+
+describe("Monasca Unit Tests", () => {
+
+    it("test", () => {
```
Looks like this is temporary code which can be removed now.

dougszumski commented on karma-test-shim.js:
```diff
@@ -0,0 +1,55 @@
+// (2)
+__karma__.loaded = function () { };
+
+// (3)
```
In general, comments should make sense to anyone reading the code. I'm not sure why these are numbered - please either explain why, or remove the comments.

dougszumski commented on karma.conf.js:
```diff
@@ -10,20 +10,33 @@ module.exports = function(config) {
     // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
-    frameworks: ['jasmine'],
+    frameworks: [ 
+      'jasmine'
```
Please move this to the reformatting commit.

dougszumski commented on src/spec/monasca_client.spec.js:
```diff
@@ -1,14 +0,0 @@
-import {MonascaClient} from "../src/components/monasca_client"
-
-describe("Monasca Unit Tests", () => {
```
It looks like this belongs in the previous commit where it was added (since we want to remove it)

dougszumski commented on src/spec/monasca_client_spec.js:
```diff
@@ -0,0 +1,12 @@
+import {MonascaClient} from "../components/monasca_client.js"
+
+describe('Something abstract', function() {
+    describe('Something specific', function() {
```
This looks like temporary test code? We don't want to merge it into master.

dougszumski commented on src/components/monasca_client.js:
```diff
@@ -19,207 +19,148 @@ export default class MonascaClient {
-          }
-
-        }
+    params.metric_dimensions = query_parameters.metric_dimensions ? query_parameters.metric_dimensions.join(",") : undefined;
```
This is much neater than the original code! Good job.

dougszumski commented on src/components/monasca_client.js:
```diff
@@ -19,207 +19,148 @@ export default class MonascaClient {
   }
 
   // Notification Method Types
-
```
More reformatting..needs to be moved to reformat commit.

dougszumski commented on src/components/monasca_client.js:
```diff
@@ -19,207 +19,148 @@ export default class MonascaClient {
 
   getAlarmHistory(id){
-    return this._get('/v2.0/alarms/' + id + '/state-history')
+    return (id == undefined ? Promise.reject("no id given to alarm history get request") : Promise.resolve(id))
```
Good stuff - validating the id before making the call is a good improvement.

dougszumski commented on src/components/monasca_client.js:
```diff
@@ -19,207 +19,148 @@ export default class MonascaClient {
 
   _patch(path, data) {
-    return this._request('PATCH', path, null, data);
+    return this._request('PATCH', path, undefined, data);
```
Please explain why we're changing this from null.

dougszumski commented on src/components/monasca_client.js:
```diff
@@ -271,18 +212,19 @@ export default class MonascaClient {
-        else {
-          throw { message: 'Monasca Error Status: ' + err.status };
+          if (monasca_response) {
+            throw { message: 'Monasca Error Response: ' + monasca_response };
```
This needs to go to the reformatting commit..

dougszumski reviewed (changes requested):
Great work adding the tests Charana - it's especially nice to see them running in Travis. After you've addressed the comments I'll go through them. Thanks.

Charana123 committed:
integrated with karma (test runner for jasmine)

Integration with karma, a javascript test runner, requiring extra
modifications to package.json to include jasmine, systemjs and karma packages, additionally,

Gruntfile.js
	- manual load of grunt plugins, to more easily identify which plugins are included
	- removed for-of babel transformer plugin because it was redundant (for-of es6
	syntax was never used)
package.json
	- removes unused packages `grunt-contrib-uglify`, `grunt-contrib-watch`,
	`babel-plugin-transform-es2015-for-of`

Charana123 committed:
systemjs loaded into karama environment

Created a test runner script `karma-test-shim.js` to locate and run spec (test) files.
Updated karma configuration script to serve required files for testing
such as source, test, systemJS (and its configuration script `sytemjs.config.js`)
and the test runner `karma-test-shim.js`

dougszumski commented on src/spec/monasca_client/notification_methods_spec.js:
```diff
@@ -0,0 +1,344 @@
+                        elements: [
+                            { type: "EMAIL" },
+                            { type: "PAGERDUTY" },
+                            { type: "WEBHOOK"}
```
We can add Slack etc in a different commit..

dougszumski reviewed (approved):
I think this is looking pretty good in general. It would be nice to squash the commits into three commits. One for adding the tests, travis and the framework, the second for reformatting commits and the third for the changes to the source code. Don't worry if it's going to cause you a lot of conflicts.

dougszumski commented on src/components/alarms.js:
```diff
@@ -100,9 +95,9 @@ export class AlarmsPageCtrl {
-    var filter = this.metricFilters[index];
-    this.metricFilters.splice(index, 1);
+  removeMetricFilter (index) {
+    // var filter = this.metricFilters[index]
```
nit: We should remove this if we don't need it

dougszumski commented on src/components/alarms.js:
```diff
@@ -122,9 +117,9 @@ export class AlarmsPageCtrl {
-    var filter = this.severityFilters[index];
-    this.severityFilters.splice(index, 1);
+  removeSeverityFilter (index) {
+    // var filter = this.severityFilters[index]
```
nit: We should remove this if we don't need it

dougszumski commented on src/components/alarms.js:
```diff
@@ -111,9 +106,9 @@ export class AlarmsPageCtrl {
-    var filter = this.stateFilters[index];
-    this.stateFilters.splice(index, 1);
+  removeStateFilter (index) {
+    // var filter = this.stateFilters[index]
```
nit: We should remove this if we don't need it

dougszumski reviewed (commented):
Please can we squash the commits?

Charana123 committed:
configured travis

Integration with travis for continous integration,
required modifications include,
karma.config.js -
	- add proxy to redirect all `base` directory file requests to the current directory,
	- use a headless browser ChromeHeadless (alternative to PhantomJS)
package.json -
	- npm test can now be used to invoke karma test runner
	- updates grunt-sass package, as travis finds difficulties importing older versions
	of node-sass (a transitive dependency of grunt-sass)

Charana123 committed:
included an trailing slash to uris indicating request to resource group/collection as opposed to a specific resource

Charana123 committed:
simplifies generation of resource request query parameters for the `listAlarms` function

Charana123 committed:
includes parameter validation (when necessary) to monasca api requests

Charana123 committed:
finished writing unit tests

Unit tests for the `MonascaClient` service

dougszumski commented on src/components/alarm_definitions.js:
```diff
@@ -22,9 +22,9 @@ import MonascaClient from './monasca_client';
 
   /** @ngInject */
-  constructor($scope, $injector, backendSrv, datasourceSrv, alertSrv) {
+  constructor($scope, $injector, backendSrv, datasourceSrv, alertSrv, monascaClientSrv) {
```
nit: Since this change makes backendSrv, datasourceSrv redundant it would be better to remove them in this commit rather than the next one.

dougszumski commented on src/components/monasca_client.js:
```diff
@@ -101,60 +96,52 @@ export default class MonascaClient {
     return this._delete('/v2.0/alarms/' + id)
       .then(resp => null)
-      .catch(err => { throw err; });
+      
```
nit: extra space
